### PR TITLE
NCAR to main candidate branch (2023-3-02)

### PR DESCRIPTION
**Pull request by alperaltuntas**
This PR includes the following bug fixes and features:
 - [add ML tracer to ideal age module](https://github.com/NCAR/MOM6/pull/222)
 - [Improvements in the gradient Ri # calculation/diagnosis](https://github.com/NCAR/MOM6/pull/225)
 - [Fix string order in the instructions for defining regional_section](https://github.com/NCAR/MOM6/pull/226)
 - [logging styling changes in nuopc cap.](https://github.com/NCAR/MOM6/pull/231)
 - [Fix NUOPC gnu restart issue by removing leading spaces in restartfiles list](https://github.com/NCAR/MOM6/pull/234)

New parameters:
`DO_BL_RESIDENCE`
`USE_REAL_BL_DEPTH`
`N_SMOOTH_RI` (previously `SMOOTH_RI`)

### CODE REVIEW

File: src/tracer/ideal_age_example.F90
Comment (Hallberg-NOAA) [inline src/tracer/ideal_age_example.F90:483]: Please indent this do-loop block by 2 points relative to the surrounding if block, to adhere to the MOM6 code standards at github.com/mom-ocean/MOM6/wiki/Code-style-guide.

File: src/tracer/ideal_age_example.F90
Comment (Hallberg-NOAA) [inline src/tracer/ideal_age_example.F90:412]: Please indent the if blocks at lines 412-416 and lines 431-434 by 2 points relative to the surrounding do loop blocks, to adhere to the MOM6 code standards at github.com/mom-ocean/MOM6/wiki/Code-style-guide (and as is already being done for other if-test blocks in this same subroutine).

File: src/tracer/ideal_age_example.F90
Comment (Hallberg-NOAA) [inline src/tracer/ideal_age_example.F90:402]: Why is layer_frac set at line 402, and then set again to the hard-coded value of 0.9 on line 403?  Is this a bug? Should one of these lines be commented out?

If both lines are intended to be there, please add a comment explaining this unusual construct.

File: src/tracer/ideal_age_example.F90
Comment (Hallberg-NOAA) [inline src/tracer/ideal_age_example.F90:382]: According to the MOM6 style guide at https://github.com/mom-ocean/MOM6/wiki/Code-style-guide, "Continuation lines should be indented at least 4 spaces, but more space can be used if it helps align lines".  Line 382 is a continuation line with a 2-point indent.  Please increase this to at least 4-spaces.

File: src/parameterizations/vertical/MOM_CVMix_shear.F90
Comment (Hallberg-NOAA) [inline src/parameterizations/vertical/MOM_CVMix_shear.F90:277]: SMOOTH_RI should be formally obsoleted by adding a line to MOM_obsolete_params.F90, so that if a user tries to use the older parameter, they get a fatal error or warning (depending on whether they have set SMOOTH_RI = True or False) with a hint suggesting that they use N_SMOOTH_RI instead.  Our standard practice is that we will keep this warning around for a few years, and then we can delete it once we have some confidence that no users are likely to accidentally use the older parameter. We don't want people accidentally (and silently) running things that they do not intend to!

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
I have visually inspected all of the changes in this PR, and apart from a handful of minor inconsistencies with the MOM6 style guide and one possible bug in ideal_age_example.F90 and a missing `obsolete_logical()` call for "SMOOTH_RI" in MOM_obsolete_params.F90 (see the specific comments for details), these changes all look good to me.  Moreover, these changes at least partially resolve several outstanding issues, including https://github.com/mom-ocean/MOM6/issues/1578 and https://github.com/NCAR/MOM6/issues/230.

  The specific changes I am requesting should be very easy to address.  We still need to run them through the GFDL pipeline testing (which we will trigger after the possible bug on lines 402 and 403 of ideal_age_example.F90 is addressed and an `obsolete_logical()` call is added for SMOOTH_RI) before formal approval from GFDL, but I would be very surprised if this turned up anything.

**Comment by jiandewang** [comment]
@alperaltuntas when you do the modification of code, can you help me remove the leading white space at
https://github.com/mom-ocean/MOM6/blob/main/config_src/drivers/nuopc_cap/mom_cap.F90#L1702-L1705
this is a leftover from my previous PR. Thanks in advance.

File: src/tracer/ideal_age_example.F90
Comment (alperaltuntas) [reply src/tracer/ideal_age_example.F90:402]: Thanks, @Hallberg-NOAA. We are waiting for a response from @sdbachman on this.

**Comment by alperaltuntas** [comment]
> @alperaltuntas when you do the modification of code, can you help me remove the leading white space at https://github.com/mom-ocean/MOM6/blob/main/config_src/drivers/nuopc_cap/mom_cap.F90#L1702-L1705 this is a leftover from my previous PR. Thanks in advance.

Done, thanks.

**Review by Hallberg-NOAA — DISMISSED** [review]

**Comment by Hallberg-NOAA** [comment]
Thank you @alperaltuntas, all of my comments on the previous version have been nicely addressed. We still need to run our pipeline regression tests (and at the moment our computer is not in a good way, so this might take some doing), but I expect that we will be giving our approval form GFDL after these tests have passed.

**Comment by jiandewang** [comment]
we got bitwise identical results in UFS, and John Stephen confirmed the same thing in N. Atlantic regional setting.

**Review by jiandewang — APPROVED** [review]
I will be on annual leave for a while starting tomorrow, so here I click the approval before GFDL's pipeline test

**Review by kshedstrom — APPROVED** [review]
I approve these changes.

File: config_src/drivers/nuopc_cap/mom_cap.F90
Comment (sanAkel) [inline config_src/drivers/nuopc_cap/mom_cap.F90:32]: Can't this `#ifdef` be replaced with a more _elegant_ `if (cesm_coupled)` that seems to work elsewhere in this module?

Comment (alperaltuntas) [reply config_src/drivers/nuopc_cap/mom_cap.F90:32]: `shr_log_mod` module is available in CESM only. So, unfortunately, an `ifdef` is the only option here.

Comment (marshallward) [reply config_src/drivers/nuopc_cap/mom_cap.F90:32]: @sanAkel Did you want to follow up?  We cannot merge until all discussions have been resolved, even if the PR has been approved.

Note that driver maintainers are largely self-maintained and generally have final say on the code.

Comment (marshallward) [reply config_src/drivers/nuopc_cap/mom_cap.F90:32]: @sanAkel I am going to close this, since the PR was approved.

There's probably a useful discussion here about when and how to preprocess, but wen can defer it to some other time.

Comment (sanAkel) [reply config_src/drivers/nuopc_cap/mom_cap.F90:32]: @marshallward, based on @alperaltuntas's answer, no objections to merging this PR.

**Pull request merged**
